### DISCUSSION TITLE
[Proposal]: Use a variable as a type

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6810] [Language Ideas]
I am curious if it would be helpful to use a variable declared as `System.Type` where the language expects a `type`.

```csharp
        // simple examples with null checks removed for clarity
        public override Boolean Equals(object? other)
        {
            Type thisType = this.GetType();

            // Compiler Error CS0118
            // thisType is a variable but is used like a type
            dynamic v = (thisType)other;

            // Compiler Error CS0118
            // thisType is a variable but is used like a type
            if (other is thisType otherIsThisType)
            {
                return Value.Equals(otherIsThisType.Value);
            }
            return false;
        }
```

Offered for discussion.

EDIT: fixed typo in condition statement.

**Comment by HaloFour** [comment]
What exactly would be the type of `v` or `otherIsThisType`?  C# is a statically-typed language and those types would need to be known at compile-time.  This suggest feels much more like something you'd see in a dynamic language where the code would be interpreted at runtime.

**Comment by ghost** [comment]
Indeed; this is a capability that resolves at runtime; and I assume negatively impact performance.

C# already allows runtime interpretation and I do not see this changing the statically-typed feature of the language. It remains explicit, if unknown.

This question asks about a way to explicitly handle circumstances when the type is not known at compile (already implemented in the language) AND the language requires providing a `type` in the syntax (e.g., casting, type conversions, generic type declarations, etc.).

```csharp
    System.Type thisType = this.GetType();
    List<thisType> foo = new();
```

Changed `var` to `dynamic`. Dynamic variables resolve at runtime.

```csharp
    System.Type thisType = this.GetType();

    // Compiler Error CS0118
    // thisType is a variable but is used like a type
    dynamic v = (thisType)other;
```

My example uses the `is` operator with type testing, and with pattern matching.

Fixed a typo in the condition.

`thisType` is defined as `System.Type`.

`otherIsThisType` is declared of type `thisType` ... which is variable.

```csharp
    // simple examples with null checks removed for clarity
    public override Boolean Equals(object? other)
    {
        System.Type thisType = this.GetType();

        // Compiler Error CS0118
        // thisType is a variable but is used like a type
        if (other is thisType otherIsThisType)
        {
            return Value.Equals(otherIsThisType.Value);
        }
        return false;
    }
```

I hope by breaking up the examples, this better illustrates the idea.

**Reply by HaloFour** [reply]
You can already perform these kinds of tasks through the use of `dynamic`, which allows you to explicitly opt-in to runtime interpretation.

**Reply by ghost** [reply]
If you can use a dynamic variable in the place of a type in the syntax, I would love to see an example.

**Reply by HaloFour** [reply]
There's no reason to cast in a dynamic type system.  If you want to do a type check you can use the `Type.IsInstanceOfType` helper method.

**Comment by Richiban** [comment]
The equivalent of Typescript's `typeof` operator would be useful, where the statically known type of variable X can be used in a type position, like so:

```cs
private read-only string data;
public void M(typeof<data> x) {} // X is of type string
```

**Reply by svick** [reply]
That has been previously suggested at https://github.com/dotnet/csharplang/discussions/1294.
